### PR TITLE
Improve Speed, Security, and Recovery

### PR DESCRIPTION
**Pull request by robinm8**
# Many Improvements

In `Dockerfile`
- Fix typo

In `production.rb`
- Compress Javascripts and CSS

In `puma.rb`
- Fix spacing

In `secrets.yml`
- Use environment variable `SECRET_KEY_BASE` in development, test, and production

In `docker-compose.yml`
- Use Docker to restart Nginx automatically in event of a crash or server reboot

In `nginx.conf`
- Fix spacing
- Use secure headers
- Remove server_name (Nginx uses server hostname automatically)
- Max connection limits (prevent DoS/crash caused by single ip address)
- Increase buffer sizes, faster TTFB
- Increase gzip compression level, for more compression.
- Use OCSP stapling (Users experience faster load times for secure content)

# Coming soon

- Migrate YACS Docker codebase 
from Debian Jessie (OpenSSL 1.0.1t; lacks ALPN support,  causes Chrome to downgrade; HTTP/1.1) 
to Alpine Linux (OpenSSL 1.0.2k; ALPN support, enables Chrome users to use latest HTTP/2.0)

  - TLDR: Chrome users will experience a faster YACS soon with h2 protocol support.

# Reference #200 

`privkey.pem` is the only secret private key of the server.

Steps to ensure secrecy of privkey.pem are essential because it protects all encrypted traffic between YACS users and the YACS server. This is to avoid traffic from being decrypted later by malicious actors (aka Hackers.)

A potential breach would mean the ability to gain access to all course data and all schedules created by YACS users along with possible data editing abilities. Limitations should be in place, to ensure malicious actors are not able to achieve that.

`cert.pem` contains only the public key that corresponds with the private key of the server.
`chain.pem` contains all of the intermediary Certificate Authority certificates.
`fullchain.pem` contains all of the certificates in `chain.pem` and the server public cert.

The files `chain.pem`, `fullchain.pem`, and `privkey.pem` are used in `Nginx.conf` to create a valid SSL profile.
The files `privkey.pem` and `cert.pem` are used in `puma.rb` because puma is not publicly facing.

In development, there is no need to generate an SSL chain (self signed root CA cert + self signed cert) because localhost is the origin server, so we only need to generate one certificate and use it as all three public cert files. (In the future, when we use an Nginx template, it will only be necessary to copy the certificate once, into `fullchain.cer`.)

Puma is a lightweight server for Ruby on Rails projects; Puma was already in place when I climbed on.. YACS. Puma is not public facing, so it only needs to verify that the private key corresponds to the public key; no need to verify the whole chain.

Nginx needs to do this because this reduces the overhead that the web browsers would have to do otherwise.

Because we want load time to be as fast as possible for YACS users, this will substantially reduce their wait time (TTFB) from over 60ms to less than 20ms.

We use SSL to verify the authenticity that the web server responding as yacs.cs.rpi.edu versus one that a potential malicious actor may try to make in attempts to spoof the DNS record. With SSL, that action would set off a big red alarm in the web browser because the malicious server that responds at that point, would not have the same private key that the valid YACS server has. Since every private key is universally unique and "impossible" to generate twice.

Web browsers automatically decide on a path to verify. Since YACS has two certification paths, `chain.pem` and `fullchain.pem` contain the certificates in path 2 except for YACS public key since path 2 implies all of the certificates that can be checked in the chain (path 2 contains all keys of path 1 and root certificate authority certificate of path 2).

The four certificates work together to ensure to the public web, the validity of the YACS server that serves on the yacs.cs.rpi.edu hostname.

The four public certificates in the YACS server certificate's chain of trust can be viewed here: https://dev.ssllabs.com/ssltest/analyze.html?d=yacs.cs.rpi.edu

I will do my best for the betterment of YACS.

### CODE REVIEW

**Comment by robinm8** [comment]
Thanks for the suggestion, edited the title and comment.

File: bin/yacs-generate-cert [hunk 1]
```diff
@@ -1,4 +1,4 @@
 #!/bin/bash
 cd $(dirname `readlink -f "$0"`)/../
 
-openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout nginx/ssl/yacs.key -out nginx/ssl/yacs.cer -subj "/C=US/ST=New York/L=Troy/O=RPI/OU=RCOS/CN=yacs.cs.rpi.edu"
+openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout nginx/ssl/fullchain.pem -out nginx/ssl/privkey.pem #-subj "/C=US/ST=New York/L=Troy/O=RPI/OU=RCOS/CN=yacs.cs.rpi.edu"
```
Comment (Bad-Science) [inline bin/yacs-generate-cert]: This script should do *everything* needed in terms of ssl setup ie it should create the chain, the fullchain, and the privkey

Comment (robinm8) [inline bin/yacs-generate-cert]: Ok, this is everything.

Comment (Bad-Science) [inline bin/yacs-generate-cert]: Thanks! LGTM

File: docker-compose.yml [hunk 2]
```diff
@@ -18,6 +18,7 @@ services:
   nginx:
     build: ./nginx
     image: nginx:mainline
+    restart: on-failure
```
Comment (Bad-Science) [inline docker-compose.yml]: :+1:

File: nginx/nginx.conf [hunk 3]
```diff
@@ -1,154 +1,182 @@
...
 
-    charset utf-8;
+        charset utf-8;
     
-    ssl_certificate_key /etc/nginx/ssl/yacs.key;
-    ssl_certificate 	/etc/nginx/ssl/chain.cer;
-    ssl_trusted_certificate /etc/nginx/ssl/yacs.cer;
+        ssl_certificate_key /etc/nginx/ssl/privkey.pem;
+        ssl_certificate 	/etc/nginx/ssl/fullchain.pem;
+        ssl_trusted_certificate /etc/nginx/ssl/chain.pem;
```
Comment (Bad-Science) [inline nginx/nginx.conf]: It looks like these are the same file in development. Are they the same file in production as well>

File: config/puma.rb [hunk 4]
```diff
@@ -10,8 +10,8 @@
 environment ENV['RAILS_ENV'] || 'development'
 
 ssl_bind '0.0.0.0', '3000', {
-  key: "/etc/puma/ssl/yacs.key",
-  cert: "/etc/puma/ssl/yacs.cer"
+    key: "/etc/puma/ssl/privkey.pem",
+    cert: "/etc/puma/ssl/cert.pem"
```
Comment (Bad-Science) [inline config/puma.rb]: Should `cert.pem` be `chain.pem` / `fullchain.pem`?

Comment (Bad-Science) [inline config/puma.rb]: Disregard this comment

**Pull request merged**